### PR TITLE
Turn on Typescript strict option for Firestore

### PR DESCRIPTION
**Pull request by hsubox76**
Set `strict: true` in Typescript compiler options.  This required setting a lot of initial values for variables  or promising the compiler (with `!`) that these values would be set before being used, so please check I have guessed the correct values or that the guarantees are correct.

### CODE REVIEW

File: packages/firestore/src/api/credentials.ts [hunk 1]
```diff
@@ -134,7 +134,7 @@ export class FirebaseCredentialsProvider implements CredentialsProvider {
   private tokenListener: ((token: string | null) => void) | null = null;
 
   /** Tracks the current User. */
-  private currentUser: User;
+  private currentUser?: User;
```
Comment (mikelehen) [inline packages/firestore/src/api/credentials.ts]: I think this should be initialized to `User.UNAUTHENTICATED` instead.

File: packages/firestore/src/api/credentials.ts [hunk 2]
```diff
@@ -192,7 +192,7 @@ export class FirebaseCredentialsProvider implements CredentialsProvider {
               typeof tokenData.accessToken === 'string',
               'Invalid tokenData returned from getToken():' + tokenData
             );
-            return new OAuthToken(tokenData.accessToken, this.currentUser);
+            return new OAuthToken(tokenData.accessToken, this.currentUser!);
```
Comment (mikelehen) [inline packages/firestore/src/api/credentials.ts]: Remove `!` after above suggestion.

**Comment by hsubox76** [comment]
This isn't urgent so I can definitely put this on the back burner until someone has time to properly review it.

**Comment by mikelehen** [comment]
I've addressed the strict violations and enabled strict mode via https://github.com/firebase/firebase-js-sdk/pull/2078.  Thanks!

**Pull request closed without merging**